src/modules/linux-micro/automount/automount.c

```diff
@@ -0,0 +1,347 @@
+    { NULL }
+};
+
+static const struct filesystem_magic ext_magic[] = {
+    { .magic = EXTMAGIC, .length = sizeof(EXTMAGIC), .offset = EXT_MAGIC_OFFSET },
+    { NULL }
+};
+
+static bool
+ext_read_superblock(int fd, char *buff)
+{
+    int err;
+
+    err = lseek(fd, EXT_FEATURE_OFFSET, SEEK_SET);
+    if (err < 0) {
+        SOL_ERR("Coud not lseek to ext superblock");
+        return false;
+    }
+
+    err = read(fd, buff, sizeof(struct ext_super_block));
+    if (err != sizeof(struct ext_super_block)) {
+        SOL_ERR("Could not read ext superblock features");
+        return false;
+    }
+
+    return true;
+}
+
+static bool
+ext2_probe_cb(int fd)
+{
+    char buff[512];
+    struct ext_super_block *sb;
+
+    if (!ext_read_superblock(fd, buff))
+        return false;
+
+    sb = (struct ext_super_block *)buff;
+    if (sb->s_feature_compat & EXT3_FEATURE_COMPAT_HAS_JOURNAL)
+        return false;
+
+    if ((sb->s_feature_ro_compat & EXT2_FEATURE_RO_COMPAT_UNSUPPORTED) ||
+        (sb->s_feature_incompat  & EXT2_FEATURE_INCOMPAT_UNSUPPORTED))
+        return false;
+
+    return true;
+}
+
+static bool
+ext3_probe_cb(int fd)
+{
+    char buff[512];
+    struct ext_super_block *sb;
+
+    if (!ext_read_superblock(fd, buff))
+        return false;
+
+    sb = (struct ext_super_block *)buff;
+    if (!(sb->s_feature_compat & EXT3_FEATURE_COMPAT_HAS_JOURNAL))
+        return false;
+
+    if ((sb->s_feature_ro_compat & EXT3_FEATURE_RO_COMPAT_UNSUPPORTED) ||
+        (sb->s_feature_incompat & EXT3_FEATURE_INCOMPAT_UNSUPPORTED))
+        return false;
+
+    return true;
+}
+
+static bool
+ext4_probe_cb(int fd)
+{
+    char buff[512];
+    struct ext_super_block *sb;
+
+    if (!ext_read_superblock(fd, buff))
+        return false;
+
+    sb = (struct ext_super_block *)buff;
+    if (sb->s_feature_incompat & EXT3_FEATURE_INCOMPAT_JOURNAL_DEV)
+        return false;
+
+    return true;
+}
+
+static const struct filesystem table[] = {
+    { .id = "vfat", .magics = vfat_magic },
+    { .id = "hfsplus", .magics = hfsplus_magic },
+    { .id = "hfs", .magics = hfs_magic },
+    { .id = "ext2", .magics = ext_magic, .fs_cb = ext2_probe_cb },
+    { .id = "ext3", .magics = ext_magic, .fs_cb = ext3_probe_cb },
+    { .id = "ext4", .magics = ext_magic, .fs_cb = ext4_probe_cb },
+};
+
+static bool
+automount_test_magic(const struct filesystem *fs, const struct filesystem_magic *magic, char *dev)
+{
+    int fd, err;
+    char buff[magic->length];
+    bool res;
+
+    fd = open(dev, O_RDONLY | O_CLOEXEC);
+    if (fd < 0) {
+        SOL_ERR("Could not open dev: %s", dev);
+        return false;
+    }
+
+    err = lseek(fd, magic->offset, SEEK_SET);
+    if (err < 0) {
+        SOL_ERR("Coud not lseek dev: %s", dev);
+        goto err;
+    }
+
+    err = read(fd, buff, magic->length);
+    if (err != magic->length) {
+        SOL_ERR("Could not read dev: %s", dev);
+        goto err;
+    }
+
+    res = streq(buff, magic->magic);
```

Comment:
Better to use `memcmp()` here, as the magic is possibly some arbitrary binary value.